fluent/context.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import absolute_import, unicode_literals
+
+import babel
+import babel.numbers
+import babel.plural
+import six
+
+from .builtins import BUILTINS
+from .resolver import resolve
+from .syntax import FluentParser
+from .syntax.ast import Message, Term
+from .utils import cachedproperty
+
+
+class MessageContext(object):
+    """
+    Message contexts are single-language stores of translations.  They are
+    responsible for parsing translation resources in the Fluent syntax and can
+    format translation units (entities) to strings.
+
+    Always use `MessageContext.format` to retrieve translation units from
+    a context.  Translations can contain references to other entities or
+    external arguments, conditional logic in form of select expressions, traits
+    which describe their grammatical features, and can use Fluent builtins.
+    See the documentation of the Fluent syntax for more information.
+    """
+
+    def __init__(self, locales, functions=None, use_isolating=True):
+        self.locales = locales
+        _functions = BUILTINS.copy()
+        if functions:
+            _functions.update(functions)
+        self._functions = _functions
+        self._use_isolating = use_isolating
+        self._messages = {}
+        self._terms = {}
+
+    def add_messages(self, source):
+        parser = FluentParser()
+        resource = parser.parse(source)
+        # TODO - warn if items get overwritten
+        for item in resource.body:
+            if isinstance(item, Message):
+                self._messages[item.id.name] = item
+            elif isinstance(item, Term):
+                self._terms[item.id.name] = item
+
+    def has_message(self, message_id):
+        try:
+            self._get_message(message_id)
+            return True
+        except LookupError:
+            return False
```

Comment:
I'm not an expert in python, but wouldn't it be faster if the `_get_message` returned `False` over throwing? Or are both optimizable comparibly in Python3?

Reply:
I haven't timed it, but:

1. This is the kind of optimization I wouldn't usually try to second guess on Python, especially now we have PyPy to think about, which could behave quite differently (and variably depending on real life context). I suspect the PyPy JIT is helped when the non-exception case always returns the same type of object.
2. In general in Python it would be considered slightly un-idiomatic to return a sentinel for failure, because then you have to always remember to handle the sentinel, and if you don't you get a more confusing error somewhere later.
3. So, for example, other bits of the code rely on `_get_message` raising an exception (e.g. `format` implicitly relies on it, it doesn't attempt to catch the exception, and this is part of its contract).